OSDOCS#6723: Included user-defined tags and labels for GCP (#63253)
openshift/openshift-docs: modules/installing-gcp-user-defined-labels-and-tags.adoc

Reviewer: Throughout this file, objects in backticks are used as nouns. Please review the [API object formatting](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting) section of the docs guidelines. 

For example, most instances of "`labels`" and "`tags`" should just be "labels" and "tags". That's the most common example here, but **no** items in backticks should be used as a plain noun. 

When possible, substitute the plain English name for a thing. When you need to discuss it as an object or field specifically, use it as an adjective and follow it with a plain English nouns (e.g., "`platform.gcp.userTags` field" as you have used in here already :slightly_smiling_face:). Plural marking should go on the plain noun.

Author: Acknowledged.

Author: Changed accordingly.

====

.User-defined labels

User-defined `labels` and {product-title} specific labels are applied only to resources created by {product-title} installer and its core operators such as: 
* `machine-api-provider-gcp` 
* `cluster-image-registry-operator`
* `gcp-filestore-csi-driver-operator`
* `gcp-pd-csi-driver-operator` 

The user-defined `labels` and {product-title} specific labels are not applied on the resources created by any other operators or the kubernetes in-tree components that create resources, for example, the Ingress load balancers.

The following are the requirements for the user-defined labels:

* A label key and value must have a minimum of 1 character and can have a maximum of 63 characters.
* A label key and value must contain only lowercase letters, numeric characters, underscore (`_`), and dash (`-`).
* A label key must start with a lowercase letter.
* You can configure a maximum of 32 labels per resource. Each resource can have a maximum of 64 labels, and 32 labels are reserved for internal use by {product-title}.

.Limitations for user-defined labels

> * Labels for `Compute Address` are supported in the GCP beta version. {product-title} does not add labels to the resource.

Reviewer 2: This is not fixed yet. Please remove space or make it lower case removing ``.

Author: Ack.

> User-defined labels and {product-title} labels are seen on the following GCP resources:

Reviewer 2: Are they seen? there could be better wordings like `available`.

Author: Since what SME meant was the labels would be found added to the GCP resources, I went for the easier alternative.

Author: changed.

* Compute Instance
* Compute Disk
* Compute Forwarding Rule
* Compute Image
* DNS Managed Zone
* Filestore Instance

> * Storage Bucket

Reviewer 2: trailing words to start with lowercase

Author: Done.

.User-defined tags

> User-defined `tags` are attached to resources created by {product-title} core operator `cluster-image-registry-operator` and not on the resources created by any other operators or the kubernetes in-tree components. In {product-title} version {product-version}, resources created by the foloowing operators are not tagged:

Reviewer 2: Also change the Operator name as suggested by Jeana.

```suggestion
User-defined tags are attached to resources created by {product-title} core operator `cluster-image-registry-operator` and not on the resources created by any other operators or the kubernetes in-tree components. In {product-title} version {product-version}, resources created by the following operators are not tagged:
```

Author: The operators name are changed as per Jeana's suggestion. Earlier they were mentioned as `machine-api-provider-gcp`, etc.

Reviewer: I think Shuba is referring to `cluster-image-registry-operator` here:
```suggestion
User-defined tags are attached to resources created by the {product-title} Cluster Image Registry Operator and not on the resources created by any other Operators or the Kubernetes in-tree components. In {product-title} version {product-version}, resources created by the following Operators are not tagged:
```
Also, note: still seeing "tags" and "labels" in backticks throughout this PR, and the word Operator in lower case (should be capitalized).

> * Installer Operator
> * Machine API Provider GCP operator 
> * GCP Filestore CSI Driver Operator
> * `GCP PD CSI Driver Operator

Reviewer 2: Suggest to confirm the operator names.
```suggestion
* Installer Operator
* Machine API Provider GCP Operator 
* GCP Filestore CSI Driver Operator
* GCP PD CSI Driver Operator
```

Author: Will check this with the SME.

Reviewer: "Machine API provider for GCP" is **not** an Operator, I have already confirmed with the dev team that owns that component.
```suggestion
* Installer Operator
* Machine API provider for GCP 
* GCP Filestore CSI Driver Operator
* GCP PD CSI Driver Operator
```

Author: Acknowledged. Changed.

** An `OrganizationID` must consist of decimal numbers, and cannot have leading zeros.
** A `ProjectID` must be 6 to 30 characters in length, can only contain lowercase letters, numbers, and hyphens, and must start with a letter, and cannot end with a hyphen.
* A tag key and value must have a minimum of 1 character and can have a maximum of 63 characters.
* A tag key must contain only uppercase and lowercase alphanumeric characters, hyphen (`-`), underscore (`_`), and period (`.`).
* A tag value must contain only uppercase and lowercase alphanumeric characters, hyphen (`-`), underscore (`_`), period (`.`), at sign (`@`), percent sign (`%`), equals sign (`=`), plusse (`+`), colon (`:`), comma (`,`), asterisk (`*`), pound sign (`$`), ampersand (`&`), parentheses (`()`), square braces (`[]`), curly braces (`{}`), and space.
* A tag key and value must begin and end with an alphanumeric character.
* Tag value must be one of the pre-defined values for the key.
* You can configure a maximum of 50 tags.
* There should be no tag key defined with the same value as any of the existing tag keys that will be inherited from the parent resource.

//Cluster creation phase

> include::installing-gcp-cluster-creation.adoc[leveloffset=+1]

Author: Changed.

> include::installing-gcp-cluster-creation.adoc[leveloffset=+1]

Reviewer 2: I am not sure what has been changed here. But guidelines do not allow this. This is an important change that you must do.

> include::installing-gcp-cluster-creation.adoc[leveloffset=+1]

Author: Is it fine to give a reference to this module here then?

> include::installing-gcp-cluster-creation.adoc[leveloffset=+1]

Reviewer 2: Reference in module is not allowed. You can go for the other option.

.Limitations for the user-defined tags

* Tags will not be attached to the following items:
** Control plane instances and storage buckets created by the installation program
** Compute instances created by the Machine API provider for GCP
** Filestore instance resources created by the GCP Filestore CSI Driver Operator

> ** Compute disk and compute image resources created by the GCP PD CSI Driver Operator

Reviewer 2: This is a duplicate content

Author: I re-arranged the initial content but, some operators may be are similiar to the labels.

Reviewer 2: Lines 53-57 sound similar. Either they are same or overlapped. Both are related to tags. They could be placed together.

Author: Removed the redundant parts.

* Tags are not supported for buckets located in the following regions:
** `us-east2` 
** `us-east3`

> * Cluster Image Registry Operator does not throw any error but skips processing tags when the buckets are created in the tags unsupported region.

Reviewer 2: I think it's Image Registry Operator. Is there something called Cluster Image Registry Operator? can you confirm?

Author: @bharath-b-rh, can you ptal at the newly created modules?

Reviewer: @snarayan-redhat "Cluster Image Registry Operator" is official (see [Operator reference](https://docs.openshift.com/container-platform/4.13/operators/operator-reference.html#cluster-image-registry-operator_cluster-operators-ref)), but you are correct that omitting "Cluster" is a very common practice in our docs.

Reviewer 2: @jeana-redhat got that! Thanks for confirming.

Author: And feedbaack from bharath: "It's called Image Registry Operator in docs, and cluster-image-registry-operator in deployments."

Reviewer 2: My suggestion would be to keep it Image Registry Operator just for consistency across docs.

Author: Ack.

* Tags must not be restricted to particular service accounts, because operators create and use service accounts with minimal roles.
* {product-title} does not create any key and value resources of the tag.
* {product-title} specific tags are not added to any resource.

> User-defined tags are seen on the GCP resource, that is, on a storage bucket.

Reviewer 2: same comment as above for `seen`

Reviewer 2: This should suffice IMO
```suggestion
User-defined tags are seen on storage bucket.
```

Author: Ack.

//Post creation Phase

> include::installing-gcp-post-cluster-creation.adoc[leveloffset=+1]

Author: Changed

* Image registry operator
* Machine API provider for GCP

The user-defined `labels` and {product-title} specific labels are not applied on the resources created by any other operators or the kubernetes in-tree components that create resources, for example, the Ingress load balancers.

The following are the requirements for the user-defined labels:

* A label key and value must have a minimum of 1 character and can have a maximum of 63 characters.
* A label key and value must contain only lowercase letters, numeric characters, underscore (`_`), and dash (`-`).
* A label key must start with a lowercase letter.
* You can configure a maximum of 32 labels per resource. Each resource can have a maximum of 64 labels, and 32 labels are reserved for internal use by {product-title}.

.Limitations for user-defined labels

* Labels for `ComputeAddress` are supported in the GCP beta version. {product-title} does not add labels to the resource.

User-defined labels and {product-title} labels are available on the following GCP resources:

* Compute disk
* Compute instance
* Compute image
* Compute forwarding rule
* DNS managed zone
* Filestore instance
* Storage bucket

.User-defined tags

> User-defined `tags` are attached to resources created by {product-title} core operator cluster image registry operator and not on the resources created by any other operators or the kubernetes in-tree components. In {product-title} version {product-version}, resources created by the foloowing operators are not tagged:

Reviewer 3: ```suggestion
User-defined `tags` are attached to resources created by {product-title} core operator cluster image registry operator and not on the resources created by any other operators or the kubernetes in-tree components. In {product-title} version {product-version}, resources created by the following operators are not tagged:
```

Author: Ack

* GCP filestore CSI driver operator
* GCP PD CSI driver operator

> * Installer operator

Reviewer 3: ```suggestion
* Installer
```

Author: Ack.

Author: As it was part of the redundant text, removed.

//Configuring the labels and tags during the cluster creation for GCP
include::modules/installing-gcp-cluster-creation.adoc[leveloffset=+1]

.Limitations for the user-defined tags

* Tags will not be attached to the following items:
** Control plane instances and storage buckets created by the installation program
** Compute instances created by the Machine API provider for GCP
** Filestore instance resources created by the GCP filestore CSI driver operator
** Compute disk and compute image resources created by the GCP PD CSI driver operator
* Tags are not supported for buckets located in the following regions:
** `us-east2` 
** `us-east3`
* Cluster image registry operator does not throw any error but skips processing tags when the buckets are created in the tags unsupported region.
* Tags must not be restricted to particular service accounts, because operators create and use service accounts with minimal roles.
* {product-title} does not create any key and value resources of the tag.
* {product-title} specific tags are not added to any resource.

> User-defined tags are available on storage bucket.

Reviewer 3: ```suggestion
User-defined tags are available on on the following GCP resources:
* Storage Bucket.
```

Author: Ack.

//Post creation Phase
include::modules/installing-gcp-post-cluster-creation.adoc[leveloffset=+1]

[role="_additional-resources"]
.Additional resources

* For more information about identifying the `OrganizationID`, see:link:https://cloud.google.com/resource-manager/docs/creating-managing-organization#retrieving_your_organization_id[OrganizationID]

* For more information about identifying the `ProjectID`, see:link:https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects[ProjectID]

> * For more information about tags and labels, see link:https://cloud.google.com/storage/docs/tags-and-labels[Tags and labels].

Reviewer 3: ```suggestion
* For more information about labels, see link:https://cloud.google.com/resource-manager/docs/labels-overview[Labels Overview].
* For more information about tags, see link:https://cloud.google.com/resource-manager/docs/tags/tags-overview[Tags Overview].
```

Author: Ack.

// Module included in the following assemblies:
// * installing/installing_gcp/installing-gcp-customizations.adoc

:_content-type: CONCEPT
[id="installing-gcp-user-defined-labels-and-tags_{context}"]
= Configuring the user-defined labels and tags for GCP

> GCP provides tags that help to identify and organize the resources created for a specific {product-title} cluster, making them easier to manage. 

Author: Ack.

> You can define labels & tags on the GCP resources only during cluster creation that is available from {product-title} version {product-version}. User-defined labels and tags are not supported for {product-title} clusters upgraded to {product-title} version {product-version}.

Author: Ack.

> User-defined labels and {product-title} specific labels are applied only to resources created by {product-title} installer and its core operators such as `machine-api-provider-gcp`, `cluster-image-registry-operator`, `gcp-filestore-csi-driver-operator`, `gcp-pd-csi-driver-operator` and not on the resources created by any other operators or the kubernetes in-tree components that creates resources for example the ingress load balancers.

Reviewer 4: {product-title}/ installation program

![Screenshot from 2023-08-28 16-05-14](https://github.com/openshift/openshift-docs/assets/57954076/ec6e836d-2571-4f0d-993a-61c79caf0437)

I think the core operators belong in a list. Large paragraph should be broken into at least two sentences. 

ingress load balancer/Ingress loab balancer. 

![Screenshot from 2023-08-28 16-06-35](https://github.com/openshift/openshift-docs/assets/57954076/412b33e5-a4cf-4434-a55a-9b75208deb3e)

Reviewer 5: s/in-tree components that creates/in-tree components that create

Author: Ack.

User-defined tags are attached to resources created by {product-title} core operator `cluster-image-registry-operator` and not on the resources created by any other operators or the kubernetes in-tree components. In {product-title} version {product-version}, resources created by `installer`, `machine-api-provider-gcp`, `gcp-filestore-csi-driver-operator`, `gcp-pd-csi-driver-operator` are not tagged.

[NOTE]
====

> Labels & tags once defined during the install-config creation phase cannot be modified or updated with new labels & tags after cluster creation.

Reviewer 4: Is Labels & tags a unique GCP term? Maybe it should be highlighted in backticks?

Reviewer 5: ```suggestion
Labels and tags are defined during the install-config creation phase, and cannot be modified or updated with new labels & tags after cluster creation.
```

====

> Support for user-defined labels & tags for GCP resources is available in {product-title} version {product-version} as Technology Preview (TP).

Reviewer 4: The technology preview snippet is missing. See guidance on https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#indicating-technology-preview-features

Author: Right, I will add the snippet. Thank you.

> Service Account used for cluster creation must have the `TagUser` roles assigned to it, required for tags defined at both organization and project level.

Reviewer 4: I'm seeing a lot of passive constructs; can these be converted into active constructs?  For example: what uses the service account?

![Screenshot from 2023-08-28 16-12-05](https://github.com/openshift/openshift-docs/assets/57954076/4bff7392-36ce-4737-aa1d-6caf8c324e61)

Author: Ack.

> You can use `.platform.gcp.userLabels` and `platform.gcp.userTags` fields to define the list of desired labels & tags in `install-config.yaml`. `.platform.gcp.userLabels` and `.platform.gcp.userTags` fields are a yaml array of objects type.

Reviewer 4: `install-config.yaml` file.

Author: Ack.

.Sample `install-config.yaml` file
[source,yaml]
----
apiVersion: v1
featureSet: TechPreviewNoUpgrade
platform:
 gcp:
   userLabels:
   - key: <label_key>
     value: <label_value>
   userTags:

>    - parentID: <OrganizationID/ProjectID>

Reviewer 4: Callouts needed to explain the user-replaced values?

Author: Let me add the callouts and their explanation.

     key: <tag_key_short_name>
     value: <tag_value_short_name>
----

The user-defined labels have the following requirements:

* A label key and value must have a minimum of 1 character and can have a maximum of 63 characters.

> * A label key and value must contain only lowercase letters, numeric characters, underscores, and dashes (i.e. _-).

Reviewer 4: (i.e. _-). > must be causing the formatting issue. I wonder if a backslash will eliminte the formatting issue?

See https://docs.asciidoctor.org/asciidoc/latest/subs/prevent/

Reviewer 5: Can probably remove the _i.e._ We are not supposed to use _i.e._ (per ISG) and the suggested replacement seems awkward here _(that is -)_.

Author: Ack.

Reviewer 4: I think (_ or -) would look better in parentheses.

Author: Done

* A label key must start with a lowercase letter.
* You can configure a maximum of 32 labels per resource.

> and each resource can have a maximum of 64 labels, and 32 labels are reserved for internal use by {product-title}.

Reviewer 6: > You can configure a maximum of 32 labels per resource. and each resource can have a maximum of 64 labels, and 32 labels are reserved for internal use by OpenShift Container Platform.

**Suggestion:** 
You can configure a maximum of 32 labels per resource. Each resource can have a maximum of 64 labels, and 32 labels are reserved for internal use by OpenShift Container Platform.

Author: Ack.

The user-defined tags have the following requirements:

> * Tag Keys and Tag Values must already exist and will not be created by {product-title}.

Reviewer 4: I'm not sure why these listed items are in italics. Is there a reason?

I also think a lot of these lists might be better placed in a reference module. I think the bulleted lists break the conceptual portrayal of information.

Author: For some reason, they turned to italics. Let me go through it again.

* A tag `parentID` can be either `OrganizationID` or `ProjectID`.
** An `OrganizationID` must consist of decimal numbers, and cannot have leading zeros.
** A `ProjectID` must be 6 to 30 characters in length, can only contain lowercase letters, numbers, and hyphens, and must start with a letter, and cannot end with a hyphen.
* A tag key and value must have a minimum of 1 character and can have a maximum of 63 characters.

> * A tag key must contain only uppercase and lowercase alphanumeric characters, hyphens, underscores, and periods (i.e. ._-).

Reviewer 6: Suggest to order the special characters by the same order of the words, for example, if the words are "hyphens, underscores, and periods", the special characters shall be listed as "-_.", thanks!

* A tag value must contain only uppercase and lowercase alphanumeric characters, hyphens, underscores, periods, at signs, percent signs, equals signs, plusses, colons, commas, asterisks, pound signs, ampersands, parentheses, square braces, curly braces, and spaces (i.e. _-.@%=+:,*#&(){}[] and spaces).

> * A tag key and value must begin and end with an alphanumeric character.

Reviewer 6: Similar suggestion as the above, as for the order of the listed special characters.

* Tag value is of type scalar and must be one of the pre-defined values for the key.
* You can configure a maximum of 50 tags.
* There should be no tag key with the same value as any of the existing tag keys on a resource(inherited from parent resource).

> For more information about identifying the `OrganizationID` or `ProjectID`, see:

Reviewer 4: Move these links to an additional resources section? In the preview, I see floating asterisks.

The lists and the links just seem dropped in among the surrounding conceptual information. 

![Screenshot from 2023-08-28 16-22-33](https://github.com/openshift/openshift-docs/assets/57954076/0a4781fa-42df-4d96-b89d-3f9f47774a99)

Reviewer 5: All of these requirements and restrictions, are they defined by Google or Red Hat? If Google, how will these lists be maintained? Would it make sense to refer users to the Google docs for these requirements and restrictions?

Author: There are many instances, that I have come across, where writers have provided links outside of Redhat. Let me get this confirmed once again.

Reviewer 4: Yeah, this is a tough area and installation docs often refer to vendor documentation as they own the platform and its capabilities.

Reviewer 4: Subheadings are great and really divide the information. Nice!

* link:https://cloud.google.com/resource-manager/docs/creating-managing-organization#retrieving_your_organization_id[OrganizationID]
* link:https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects[ProjectID]

> After creating {product-title} cluster, you can access the list of the labels and tags defined for the GCP resources by referring to the new fields `.status.platformStatus.gcp.resourceLabels` and `.status.platformStatus.gcp.resourceTags` present in `infrastructures.config.openshift.io/cluster` object.

Reviewer 4: ```suggestion
After creating {product-title} cluster, you can access the list of the labels and tags defined for the GCP resources by referring to the new fields `.status.platformStatus.gcp.resourceLabels` and `.status.platformStatus.gcp.resourceTags` present in the `infrastructures.config.openshift.io/cluster` object.
```

.Sample `infrastructure` file
[source,yaml]
----
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
 name: cluster
spec:
 platformSpec:
   type: GCP
status:
 infrastructureName: <cluster_id>
 platform: GCP
 platformStatus:
   gcp:
     resourceLabels:
     - key: <label_key>
       value: <label_value>
     resourceTags:
     - key: <tag_key_short_name>
       parentID: <OrganizationID/ProjectID>
       value: <tag_value_short_name>
   type: GCP
----

Along with the user-defined labels, resources will also have a label defined by the {product-title}. The format of the {product-title} labels is `kubernetes-io-cluster-<cluster_id>:owned`. The `cluster_id` parameter is the value of `.status.infrastructureName` present in `infrastructures.config.openshift.io/cluster`.

User-defined and {product-title} labels can be found on the following GCP resoucres:

* Compute Instance
* Compute Disk
* Compute Forwarding Rule
* Compute Image
* DNS Managed Zone
* Filestore Instance
* Storage Bucket

User-defined tags can be found added to the GCP resources, that is, storage bucket.

> The following are the limitations for the user-defined tags:

Reviewer 4: IMO, this should be a new sub-module.

Author: @dfitzmau Do we need to create submodules for both labels and tags?

Reviewer 4: I think one reference module would be enough. This module could capture the folowing lists:

> The user-defined labels have the following requirements:
> The user-defined tags have the following requirements:
> The following are the limitations for the user-defined tags:

I'll leave this choice up to you. I personally found the multiple lists in the concept module difficult to follow, but maybe the limitations list could be stripped out and placed in a reference module or even a separate concept module. Totally up to you.

Reviewer 4: No longer required as sub-heading work really well.

Reviewer 2: +1 @dfitzmau

> * Tags will not be attached to master compute instances and storage bucket created by `installer`, worker compute instances created by `machine-api-provider-gcp controller`, filesore instance resources created by `gcp-filestore-csi-driver-operator` and compute disk, compute image resources created by `gcp-pd-csi-driver-operator`.

Reviewer 5: ```suggestion
* Tags will not be attached to master compute instances and storage bucket created by `installer`, worker compute instances created by `machine-api-provider-gcp controller`, filestore instance resources created by `gcp-filestore-csi-driver-operator` and compute disk, compute image resources created by `gcp-pd-csi-driver-operator`.
```

* Tags are not supported for buckets located in the following regions:
** `us-east2` 

> **`us-east3`

Reviewer 5: This one is not rendering correctly. Maybe you need a space between the ** and `us-east3`?

![image](https://github.com/openshift/openshift-docs/assets/25209155/d7ec7bac-a823-41d8-9810-9a6bb8b62a7a)

Author: Ack.

* `cluster-image-registry-operator` will not throw any error but skips processing tags when the buckets are created in the tags unsupported region.
* Tags must not be restricted to particular service accounts, since operators create and use service accounts with minimal roles.
* {product-title} will not create tag keys and value resources.
* {product-title} specific tags will not be added to any resource.
* Labels for `Compute Address` are supported in the GCP beta version and {product-title} does not add labels to the resource.

> For more information about tags and labels, see link:https://cloud.google.com/storage/docs/tags-and-labels#tags[Tags and labels].

Author: Done

// Module included in the following assemblies:
// * installing/installing_gcp/installing-gcp-customizations.adoc

:_content-type: CONCEPT
[id="installing-gcp-user-defined-labels-and-tags_{context}"]
= Configuring the user-defined labels and tags for GCP

GCP provides tags that help to identify and organize the resources created for a specific {product-title} cluster, making them easier to manage. 

You can define labels & tags on the GCP resources only during cluster creation that is available from {product-title} version {product-version}. User-defined labels and tags are not supported for {product-title} clusters upgraded to {product-title} version {product-version}.

User-defined labels and {product-title} specific labels are applied only to resources created by {product-title} installer and its core operators such as `machine-api-provider-gcp`, `cluster-image-registry-operator`, `gcp-filestore-csi-driver-operator`, `gcp-pd-csi-driver-operator` and not on the resources created by any other operators or the kubernetes in-tree components that creates resources for example the ingress load balancers.

> User-defined tags are attached to resources created by {product-title} installer and its core operators such as `cluster-image-registry-operator` and not on the resources created by any other operators or the kubernetes in-tree components. In {product-title} version {product-version}, resources created by `machine-api-provider-gcp`, `gcp-filestore-csi-driver-operator`, `gcp-pd-csi-driver-operator` are not tagged.

Author: Ack.

[NOTE]
====
Labels & tags once defined during the install-config creation phase cannot be modified or updated with new labels & tags after cluster creation.
====

Support for user-defined labels & tags for GCP resources is available in {product-title} version {product-version} as Technology Preview (TP).

> Service Account used for cluster creation must have the `TagUser` role assigned to it, required for tags defined at both organization and project level.

Reviewer 6: FYI About the required role(s), I need to get a successful installation to double confirm. Will let you know then, thanks!

Author: No issues.

Author: @jianli-wei any updates?

> Service Account used for cluster creation must have the `TagUser` role assigned to it, required for tags defined at both organization and project level.

Reviewer 3: `TagViewer` role is also required. Will wait for @jianli-wei confirmation.

```suggestion
Service Account used for cluster creation must have the `TagViewer` and `TagUser` roles assigned to it, required for tags defined at both organization and project level.
```

Author: Ack.

Reviewer 6: @subhtk Yes, I was trying the installations using 4.14 nightly latest yesterday, unfortunately failed (see [comment](https://issues.redhat.com/browse/CORS-2536?focusedId=22849021&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-22849021)), I guess more debugging needed. 

Maybe you could move on now, and we could use another PR to correct the required roles if necessary. Thanks!

Reviewer 6: @subhtk I did some tests as below, and I think we could remove `Tag Viewer` role here. 

You can see `Tag User` role has all the permissions of `Tag Viewer` role, the only exception is `resourcemanager.tagHolds.list` permission, which has been included in the pre-defined role `Security Admin` (or IAM Security Admin), which is in turn documented already (see [OCP doc](https://docs.openshift.com/container-platform/4.13/installing/installing_gcp/installing-gcp-account.html#installation-gcp-permissions_installing-gcp-account) section "_Required roles for the installation program_").

```
$ gcloud iam roles describe roles/resourcemanager.tagUser > /tmp/resourcemanager.tagUser
$ gcloud iam roles describe roles/resourcemanager.tagViewer > /tmp/resourcemanager.tagViewer
$ diff /tmp/resourcemanager.tagUser /tmp/resourcemanager.tagViewer | grep -E "^> "
> description: Access to list Tags and their associations with resources
> - resourcemanager.tagHolds.list
> name: roles/resourcemanager.tagViewer
> title: Tag Viewer
$ gcloud iam roles describe roles/iam.securityAdmin | grep resourcemanager.tagHolds.list
- resourcemanager.tagHolds.list
$ 
```

I tried with a QE SA, after removing `Tag Viewer` role from the SA, to list tag keys/values under Org level or QE projects level still works well. @bharath-b-rh Please advise, thanks!

Reviewer 3: @jianli-wei Thank you for confirming! Yeah `TagUser` supersedes `TagViewer`, since during tests we faced some issues thought it was also required.
@subhtk Please disregard my comment to add `TagViewer` role.

Author: Okay.

Author: @bharath-b-rh Done with the changes, have a final review?

You can use `.platform.gcp.userLabels` and `platform.gcp.userTags` fields to define the list of desired labels & tags in `install-config.yaml`. `.platform.gcp.userLabels` and `.platform.gcp.userTags` fields are a yaml array of objects type.

.Sample `install-config.yaml` file
[source,yaml]
----
apiVersion: v1
featureSet: TechPreviewNoUpgrade
platform:
 gcp:
   userLabels:
   - key: <label_key>
     value: <label_value>
   userTags:
   - parentID: <OrganizationID/ProjectID>
     key: <tag_key_short_name>
     value: <tag_value_short_name>
----

The user-defined labels have the following requirements:

* A label key and value must have a minimum of 1 character and can have a maximum of 63 characters.

> * A label key and value must contain only lowercase letters, numeric characters, underscores, and dashes.

Reviewer 6: If possible, suggest to list the allowed special characters directly, e.g. 

* A label key and value must contain only lowercase letters, numeric characters, and the following special characters, including underscores, and dashes (i.e. `_-`).

Author: Ack.

* A label key must start with a lowercase letter.
* Each resource can have a maximum of 32 labels, and 32 labels are reserved for internal use by {product-title}.

> * You can configure a maximum of 64 labels per resource.

Reviewer 6: Sorry, I'm confused a bit with the above 2 statements (line 49 and line 50), e.g. how many user-defined labels can be specified at most, in `platform.gcp.userLabels` of `install-config.yaml`? Thanks!

Author: The draft content was: 

* Each resource can have multiple labels, up to a maximum of 32.

**  GCP supports a maximum of 64 labels per resource. Restricting the number of user defined labels to 32 and reserving 32 for Openshift's internal use.

I will change the content to -
* You can configure a maximum of 64 labels per resource.
** Each resource can have a maximum of 32 labels, and 32 labels are reserved for internal use by {product-title}.

Author: Seems like the meaning that needs to be delivered through these steps is:
* Each resource can have multiple labels, up to a maximum of 32.
** GCP supports a maximum of 64 labels per resource. Restricting the number of user defined labels to 32 and reserving 32 for Openshift's internal use.

So, I will have to revert the changes back.

The user-defined tags have the following requirements:

* Tag Keys and Tag Values must already exist and will not be created by {product-title}.
* A tag `parentID` can be either `OrganizationID` or `ProjectID`.
** An `OrganizationID` must consist of decimal numbers, and cannot have leading zeros.
** A `ProjectID` must be 6 to 30 characters in length, can only contain lowercase letters, numbers, and hyphens, and must start with a letter, and cannot end with a hyphen.
* A tag key and value must have a minimum of 1 character and can have a maximum of 63 characters.

> * A tag key must contain only uppercase and lowercase alphanumeric characters, hyphens, underscores, and periods.

Reviewer 6: Similarly, if possible, also list the allowed special characters directly, e.g.

* A tag key must contain only uppercase and lowercase alphanumeric characters, and the following special characters, including hyphens, underscores, and periods (i.e. `._-`).

Author: Ack.

> * A tag value must contain only uppercase and lowercase alphanumeric characters, hyphens, underscores, periods, at signs, percent signs, equals signs, plusses, colons, commas, asterisks, pound signs, ampersands, parentheses, square braces, curly braces, and spaces.

Reviewer 6: Similarly, if possible, also list the allowed special characters directly, e.g. 

* A tag value must contain only uppercase and lowercase alphanumeric characters, and the following special characters, including hyphens, underscores, periods, at signs, percent signs, equals signs, plusses, colons, commas, asterisks, pound signs, ampersands, parentheses, square braces, curly braces, and spaces (i.e. `_-.@%=+:,*#&(){}[]` and spaces).

Author: Done.

* A tag key and value must begin and end with an alphanumeric character.
* Tag value is of type scalar and must be one of the pre-defined values for the key.
* You can configure a maximum of 50 tags.
* There should be no tag key with the same value as any of the existing tag keys on a resource(inherited from parent resource).

> For more information about identifying the `OrganizationID` or `ProjectID`, see:

Reviewer 3: ```suggestion
For more information about identifying the `OrganizationID` or `ProjectID`, see:

* link:https://cloud.google.com/resource-manager/docs/creating-managing-organization#retrieving_your_organization_id[OrganizationID]
* link:https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects[ProjectID]
```

Author: Ack.

.Sample `infrastructure` file
[source,yaml]
----
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
 name: cluster
spec:
 platformSpec:
   type: GCP
status:
 infrastructureName: <cluster_id>
 platform: GCP
 platformStatus:
   gcp:
     resourceLabels:
     - key: <label_key>
       value: <label_value>
     resourceTags:
     - key: <tag_key_short_name>
       parentID: <OrganizationID/ProjectID>
       value: <tag_value_short_name>
   type: GCP
----

Along with the user-defined labels, resources will also have a label defined by the {product-title}. The format of the {product-title} labels is `kubernetes-io-cluster-<cluster_id>:owned`. The `cluster_id` parameter is the value of `.status.infrastructureName` present in `infrastructures.config.openshift.io/cluster`.

User-defined and {product-title} labels can be found on the following GCP resoucres:

* Compute Instance
* Compute Disk
* Compute Forwarding Rule
* Compute Image
* DNS Managed Zone
* Filestore Instance
* Storage Bucket

User-defined tags can be found added to the following GCP resources:

> * Compute Instance

Reviewer 3: ```suggestion
```

* Storage Bucket

The following are the limitations for the user-defined tags:

> * Tags will not be attached to worker compute instances created by `machine-api-provider-gcp controller`, filesore instance resources created by `gcp-filestore-csi-driver-operator` and compute disk, compute image resources created by `gcp-pd-csi-driver-operator`.

Author: Ack.

> * Tags are not supported for buckets located in the following regions:

Reviewer 3: ```suggestion
* Tags are not supported for buckets located in the `us-east2` and `us-east3` regions. 
** `cluster-image-registry-operator` will not throw any error but skips processing tags when the buckets are created in the tags unsupported region.
```

Author: Ack.

// Module included in the following assemblies:
// * installing/installing_gcp/installing-gcp-customizations.adoc

:_content-type: CONCEPT
[id="installing-gcp-user-defined-labels-and-tags_{context}"]
= Configuring the user-defined labels and tags for GCP

Google Cloud Platform (GCP) provides `labels` and `tags` that help to identify and organize the resources created for a specific {product-title} cluster, making them easier to manage. 

You can define `labels` and `tags` for each GCP resource only during {product-title} cluster installation operation. 

[IMPORTANT]
====

> User-defined `labels` and `tags` are not supported for {product-title} clusters upgraded to {product-title} version {product-version}.

Author: 4.14. I have changed it accordingly. Sorry for the confusion.

====

User-defined `labels` and {product-title} specific labels are applied only to resources created by {product-title} installer and its core operators such as: 
* `machine-api-provider-gcp` 
* `cluster-image-registry-operator`
* `gcp-filestore-csi-driver-operator`
* `gcp-pd-csi-driver-operator` 

The user-defined `labels` and {product-title} specific labels are not applied on the resources created by any other operators or the kubernetes in-tree components that create resources, for example, the Ingress load balancers.

User-defined `tags` are attached to resources created by {product-title} core operator `cluster-image-registry-operator` and not on the resources created by any other operators or the kubernetes in-tree components. In {product-title} version {product-version}, resources created by `installer`, `machine-api-provider-gcp`, `gcp-filestore-csi-driver-operator`, `gcp-pd-csi-driver-operator` are not tagged.

[NOTE]
====

> `Labels` and `tags` are defined during the `install-config.yaml` creation phase, and cannot be modified or updated with new `labels` & `tags` after cluster creation.

Author: Done

====

:FeatureName: Support for user-defined `labels` and `tags` for GCP
include::snippets/technology-preview.adoc[]

The installation program requires that a service account includes a `TagUser` role, so that the program can create the {product-title} cluster with defined tags at both organization and project levels.

> You can use the `.platform.gcp.userLabels` and `platform.gcp.userTags` fields to define the list of desired labels & tags in `install-config.yaml` file. The `.platform.gcp.userLabels` and `.platform.gcp.userTags` fields are a YAML array of objects type.

Author: Ack.

.Sample `install-config.yaml` file
[source,yaml]
----
apiVersion: v1
featureSet: TechPreviewNoUpgrade
platform:
 gcp:
   userLabels:<1>
   - key: <label_key><2>
     value: <label_value><3>
   userTags:<4>
   - parentID: <OrganizationID/ProjectID><5>
     key: <tag_key_short_name><6>
     value: <tag_value_short_name><7>
----
<1> `userLabels`: Adds keys and values as labels to the resources created on GCP.
<2> `key`: The label name.
<3> `value`: The label content.
<4> `userTags`: Adds keys and values as tags to the resources created on GCP.

> <5> `parentID`: The ID of the hierarchical resource where the tags are defined, at the Organization or the Project level.

Author: Ack

<6> `key`: The tag name.
<7> `value`: The tag content.

.Requirements for the user-defined labels

* A label key and value must have a minimum of 1 character and can have a maximum of 63 characters.
* A label key and value must contain only lowercase letters, numeric characters, underscores, and dashes (-).
* A label key must start with a lowercase letter.
* You can configure a maximum of 32 labels per resource. Each resource can have a maximum of 64 labels, and 32 labels are reserved for internal use by {product-title}.

.Requirements for the user-defined tags

> * Tag Keys and Tag Values must already exist and will not be created by {product-title}.

Author: Done

* A tag `parentID` can be either `OrganizationID` or `ProjectID`:
** An `OrganizationID` must consist of decimal numbers, and cannot have leading zeros.
** A `ProjectID` must be 6 to 30 characters in length, can only contain lowercase letters, numbers, and hyphens, and must start with a letter, and cannot end with a hyphen.
* A tag key and value must have a minimum of 1 character and can have a maximum of 63 characters.
* A tag key must contain only uppercase and lowercase alphanumeric characters, hyphens (-), underscores (_), and periods (.).
* A tag value must contain only uppercase and lowercase alphanumeric characters, hyphens (-), underscores (_), periods (.), at signs (@), percent signs (%), equals signs (=), plusses (+), colons (:), commas (,), asterisks (*), pound signs ($), ampersands (&), parentheses (), square braces [], curly braces {}, and spaces.
* A tag key and value must begin and end with an alphanumeric character.

> * Tag value is of type scalar and must be one of the pre-defined values for the key.

Reviewer 2: Also put all the brackets types in parantheses.

> * Tag value is of type scalar and must be one of the pre-defined values for the key.

Author: Ack.

* You can configure a maximum of 50 tags.

> * There should be no tag key with the same value as any of the existing tag keys on a resource (inherited from parent resource).

Author: Ack. How does this sound? 
"There should be no tag key with the same value as any of the existing tag keys on a resource, inherited from the parent resource."

> * There should be no tag key with the same value as any of the existing tag keys on a resource (inherited from parent resource).

Reviewer 2: `There should be no tag key with the same value as any of the existing tag keys that are inherited from the parent resource.`
If this is what it means. Please confirm the technical accuracy.

> * There should be no tag key with the same value as any of the existing tag keys on a resource (inherited from parent resource).

Author: Confirmed with SME. Suggestion from SME:- "There should be no tag key defined with the same value as any of the existing tag keys that will be inherited from the parent resource."

> .Configuration after the cluster creation

Author: Addressed.

After creating the {product-title} cluster, you can access the list of the labels and tags defined for the GCP resources by referring to the new fields `.status.platformStatus.gcp.resourceLabels` and `.status.platformStatus.gcp.resourceTags` present in the `infrastructures.config.openshift.io/cluster` object.

.Sample `infrastructure.yaml` file
[source,yaml]
----
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
 name: cluster
spec:
 platformSpec:
   type: GCP
status:
 infrastructureName: <cluster_id>
 platform: GCP
 platformStatus:
   gcp:
     resourceLabels:
     - key: <label_key>
       value: <label_value>
     resourceTags:
     - key: <tag_key_short_name>
       parentID: <OrganizationID/ProjectID>
       value: <tag_value_short_name>
   type: GCP
----

> Along with the user-defined labels, resources will also have a label defined by the {product-title}. The format of the {product-title} labels is `kubernetes-io-cluster-<cluster_id>:owned`. The `cluster_id` parameter is the value of `.status.infrastructureName` present in `infrastructures.config.openshift.io/cluster`.

Author: Ack.

> Along with the user-defined labels, resources will also have a label defined by the {product-title}. The format of the {product-title} labels is `kubernetes-io-cluster-<cluster_id>:owned`. The `cluster_id` parameter is the value of `.status.infrastructureName` present in `infrastructures.config.openshift.io/cluster`.

Author: Ack.

> User-defined and {product-title} labels can be found on the following GCP resoucres:

Author: Ack

* Compute Instance
* Compute Disk
* Compute Forwarding Rule
* Compute Image
* DNS Managed Zone
* Filestore Instance
* Storage Bucket

User-defined tags can be found added to the GCP resources, that is, storage bucket.

> .The following are the limitations for the user-defined tags:

Author: Ack.

* Tags are not attached to master compute instances or storage bucket that are created by `installer`, and worker compute instances that are created by `machine-api-provider-gcp controller`, filestore instance resources created by `gcp-filestore-csi-driver-operator` and compute disk, compute image resources created by `gcp-pd-csi-driver-operator`.
* Tags are not supported for buckets located in the following regions:
** `us-east2` 
** `us-east3`
* `cluster-image-registry-operator` does not throw any error but skips processing tags when the buckets are created in the tags unsupported region.

> * Tags must not be restricted to particular service accounts, because operators create and use service accounts with minimal roles.

Author: Ack

> * {product-title} does not create any tag keys and value resources.

Reviewer 2: I am afraid it does not. It talks about the tag's key and value.

> * {product-title} does not create any tag keys and value resources.

Author: I will change it accordingly.

// Module included in the following assemblies:
// * installing/installing_gcp/installing-gcp-customizations.adoc

:_content-type: CONCEPT
[id="installing-gcp-user-defined-labels-and-tags_{context}"]
= Configuring the user-defined labels and tags for GCP

Google Cloud Platform (GCP) provides `labels` and `tags` that help to identify and organize the resources created for a specific {product-title} cluster, making them easier to manage. 

You can define `labels` and `tags` for each GCP resource only during {product-title} cluster installation operation. 

[IMPORTANT]
====
User-defined `labels` and `tags` are not supported for {product-title} clusters upgraded to {product-title} version {product-version}.
====

User-defined `labels` and {product-title} specific labels are applied only to resources created by {product-title} installer and its core operators such as: 
* `machine-api-provider-gcp` 
* `cluster-image-registry-operator`
* `gcp-filestore-csi-driver-operator`
* `gcp-pd-csi-driver-operator` 

> And they are not applied on the resources created by any other operators or the kubernetes in-tree components that create resources for example the Ingress load balancers.

Author: Ack.

The user-defined `labels` and {product-title} specific labels are not applied on the resources created by any other operators or the kubernetes in-tree components that create resources, for example, the Ingress load balancers.

The following are the requirements for the user-defined labels:

* A label key and value must have a minimum of 1 character and can have a maximum of 63 characters.

> * A label key and value must contain only lowercase letters, numeric characters, underscore (_), and dash (-).

Reviewer: Noticed the text I was asked to look at is in italics, seems this line is why. 

Should put all the literal characters in backticks throughout this file, e.g.:
```suggestion
* A label key and value must contain only lowercase letters, numeric characters, underscore (`_`), and dash (`-`).
```

Author: This helps a lot. Thank you.

> `Labels` and `tags` are defined during the install-config creation phase, and cannot be modified or updated with new `labels` & `tags` after cluster creation.

Reviewer 4: Thanks, yes. Many [examples](https://docs.openshift.com/container-platform/4.13/installing/installing_vsphere/installing-vsphere-installer-provisioned-network-customizations.html) of this file's use in install docs. To be more precise, consider:

> `install-config.yaml` installation configuration file

> Support for user-defined `labels` & `tags` for GCP resources is available in {product-title} version {product-version} as Technology Preview (TP).

Reviewer 4: Agreed

> :FeatureName: User-defined `labels` and `tags`

Reviewer 4: lables "and" tags?

include::snippets/technology-preview.adoc[]

> The installation program requires that a service account includes a TagUser role, so that the program can create the {product-title} cluster with defined tags at both organization and project levels.

Author: Ack.

> You can use `.platform.gcp.userLabels` and `platform.gcp.userTags` fields to define the list of desired labels & tags in `install-config.yaml` file. `.platform.gcp.userLabels` and `.platform.gcp.userTags` fields are a YAML array of objects type.

Author: Ack.

.Sample `install-config.yaml` file
[source,yaml]
----
apiVersion: v1
featureSet: TechPreviewNoUpgrade
platform:
 gcp:
   userLabels:<1>
   - key: <label_key><2>
     value: <label_value><3>
   userTags:<4>
   - parentID: <OrganizationID/ProjectID><5>
     key: <tag_key_short_name><6>
     value: <tag_value_short_name><7>
----
<1> `userLabels`: userLabels has keys and values that will be added as labels to the resources created on GCP.
<2> `key`: key is the label name.
<3> `value`: value is the label content.
<4> `userTags`: userTags has keys and values that will be added as tags to the resources created on GCP.

> <5> `parentID`: parentID is the ID of the hierarchical resource where the tags are defined, e.g. at the Organization or the Project level.

Author: Ack.

<6> `key`: key is the tag name.

> <7> `value`: value is the tag content.

Author: Ack.

.Requirements for the user-defined labels

* A label key and value must have a minimum of 1 character and can have a maximum of 63 characters.
* A label key and value must contain only lowercase letters, numeric characters, underscores, and dashes (that is, _-).
* A label key must start with a lowercase letter.
* You can configure a maximum of 32 labels per resource. Each resource can have a maximum of 64 labels, and 32 labels are reserved for internal use by {product-title}.

.Requirements for the user-defined tags

* Tag Keys and Tag Values must already exist and will not be created by {product-title}.

> * A tag `parentID` can be either `OrganizationID` or `ProjectID`.

Author: Ack.

** An `OrganizationID` must consist of decimal numbers, and cannot have leading zeros.
** A `ProjectID` must be 6 to 30 characters in length, can only contain lowercase letters, numbers, and hyphens, and must start with a letter, and cannot end with a hyphen.
* A tag key and value must have a minimum of 1 character and can have a maximum of 63 characters.
* A tag key must contain only uppercase and lowercase alphanumeric characters, hyphens, underscores, and periods (that is, -_.).
* A tag value must contain only uppercase and lowercase alphanumeric characters, hyphens, underscores, periods, at signs, percent signs, equals signs, plusses, colons, commas, asterisks, pound signs, ampersands, parentheses, square braces, curly braces, and spaces (that is, -_.@%=+:,*$&()[]{} and spaces).
* A tag key and value must begin and end with an alphanumeric character.
* Tag value is of type scalar and must be one of the pre-defined values for the key.
* You can configure a maximum of 50 tags.
* There should be no tag key with the same value as any of the existing tag keys on a resource (inherited from parent resource).

> .Post-cluster creation

Author: Ack.

> After creating {product-title} cluster, you can access the list of the labels and tags defined for the GCP resources by referring to the new fields `.status.platformStatus.gcp.resourceLabels` and `.status.platformStatus.gcp.resourceTags` present in the `infrastructures.config.openshift.io/cluster` object.

Author: Ack.

> .Sample `infrastructure` file

Author: Ack.

[source,yaml]
----
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
 name: cluster
spec:
 platformSpec:
   type: GCP
status:
 infrastructureName: <cluster_id>
 platform: GCP
 platformStatus:
   gcp:
     resourceLabels:
     - key: <label_key>
       value: <label_value>
     resourceTags:
     - key: <tag_key_short_name>
       parentID: <OrganizationID/ProjectID>
       value: <tag_value_short_name>
   type: GCP
----

Along with the user-defined labels, resources will also have a label defined by the {product-title}. The format of the {product-title} labels is `kubernetes-io-cluster-<cluster_id>:owned`. The `cluster_id` parameter is the value of `.status.infrastructureName` present in `infrastructures.config.openshift.io/cluster`.

User-defined and {product-title} labels can be found on the following GCP resoucres:

* Compute Instance
* Compute Disk
* Compute Forwarding Rule
* Compute Image
* DNS Managed Zone
* Filestore Instance
* Storage Bucket

User-defined tags can be found added to the GCP resources, that is, storage bucket.

The following are the limitations for the user-defined tags:

> * Tags will not be attached to master compute instances and storage bucket created by `installer`, worker compute instances created by `machine-api-provider-gcp controller`, filestore instance resources created by `gcp-filestore-csi-driver-operator` and compute disk, compute image resources created by `gcp-pd-csi-driver-operator`.

Reviewer 2: - Remove future tense
- what is 'master compute' instances here? We generally don't use master term but depends on the context.same applies to worker.
- What is installer? If it is not an object type, we have to simply call it installer without quotes.

Reviewer 4: installer > installation program

Author: Ack.

Reviewer 4: Hi @subhtk . I'd say you can ignore this. Apologies. I see that the [Machine Management ](https://docs.openshift.com/container-platform/4.13/machine_management/creating-infrastructure-machinesets.html#binding-infra-node-workloads-using-taints-tolerations_creating-infrastructure-machinesets) docs  still use the value, so the engineering codebase must not be updating the name of this node.

Reviewer 4: After you resolve the merge conflict, this PR is ready for merging. Good work!

Reviewer 2: @subhtk we do not use term master and worker anymore. Please find this [guideline](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#using-conscious-language). As I mentioned in my earlier comment, kindly figure out what master compute instances mean and replace with the required wordings.

Reviewer 4: @jeana-redhat . Would you be OK to confirm if these `master` entries are not changeable? They seem to target engineering codebase items. As writers, we can control generic reference to components but not codebase items.

Reviewer: Short answer: no, I don't think these should be master and worker in this PR. Long answer coming in a bit, want to confirm a term with dev, please stand by

* Tags are not supported for buckets located in the following regions:
** `us-east2` 
** `us-east3`

> * `cluster-image-registry-operator` will not throw any error but skips processing tags when the buckets are created in the tags unsupported region.

Reviewer 2: future tense

Author: Ack.

> * Tags must not be restricted to particular service accounts, since operators create and use service accounts with minimal roles.

Reviewer 2: ```suggestion
* Tags must not be restricted to particular service accounts, because operators create and use service accounts with minimal roles.
```

Author: Ack.

* {product-title} will not create tag keys and value resources.
* {product-title} specific tags will not be added to any resource.

> * Labels for `Compute Address` are supported in the GCP beta version and {product-title} does not add labels to the resource.

Reviewer 2: ```suggestion
* Labels for `Compute Address` are supported in the GCP beta version. {product-title} does not add labels to the resource.
```

Author: Ack.

[role="_additional-resources"]
.Additional resources

* For more information about identifying the `OrganizationID` or `ProjectID`, see:
** link:https://cloud.google.com/resource-manager/docs/creating-managing-organization#retrieving_your_organization_id[OrganizationID]
** link:https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects[ProjectID]

> * For more information about tags and labels, see link:https://cloud.google.com/storage/docs/tags-and-labels#tags[Tags and labels].

Reviewer 3: We would need different references for tags and labels. Please include below links
Tags: https://cloud.google.com/resource-manager/docs/tags/tags-overview
Labels: https://cloud.google.com/resource-manager/docs/creating-managing-labels

Existing link https://cloud.google.com/storage/docs/tags-and-labels#tags is specific to storage buckets, it won't be required.

Author: Ack.

// Module included in the following assemblies:
// * installing/installing_gcp/installing-gcp-customizations.adoc

:_content-type: CONCEPT
[id="installing-gcp-user-defined-labels-and-tags_{context}"]
= Configuring the user-defined labels and tags for GCP

Google Cloud Platform (GCP) provides `labels` and `tags` that help to identify and organize the resources created for a specific {product-title} cluster, making them easier to manage. 

You can define `labels` and `tags` for each GCP resource only during {product-title} cluster installation. 

[IMPORTANT]
====
User-defined `labels` and `tags` are not supported for {product-title} clusters upgraded to {product-title} 4.14 version.
====

.User-defined labels

User-defined `labels` and {product-title} specific labels are applied only to resources created by {product-title} installer and its core operators such as: 
* `machine-api-provider-gcp` 
* `cluster-image-registry-operator`
* `gcp-filestore-csi-driver-operator`

> * `gcp-pd-csi-driver-operator` 

Reviewer: For this section, and any other remaining use of Operator image names in this PR, please see the last part of my comment here: https://github.com/openshift/openshift-docs/pull/63253#discussion_r1341431433

Author: I will change the operator names accordingly.

The user-defined `labels` and {product-title} specific labels are not applied on the resources created by any other operators or the kubernetes in-tree components that create resources, for example, the Ingress load balancers.

> The following are the requirements for the user-defined labels:

Reviewer 2: This should be present in a location when user is trying to create a label.

Author: I will re-arrange the info while creating the seperate modules.

Reviewer 2: Not addressed

* A label key and value must have a minimum of 1 character and can have a maximum of 63 characters.
* A label key and value must contain only lowercase letters, numeric characters, underscore (`_`), and dash (`-`).
* A label key must start with a lowercase letter.
* You can configure a maximum of 32 labels per resource. Each resource can have a maximum of 64 labels, and 32 labels are reserved for internal use by {product-title}.

.User-defined tags

User-defined `tags` are attached to resources created by {product-title} core operator `cluster-image-registry-operator` and not on the resources created by any other operators or the kubernetes in-tree components. In {product-title} version {product-version}, resources created by `installer`, `machine-api-provider-gcp`, `gcp-filestore-csi-driver-operator`, `gcp-pd-csi-driver-operator` are not tagged.

> The following are the requirements for the user-defined tags:

Reviewer 2: This should be present when user is trying to create a tag.

Author: Ok

Author: Here, since I have created two sub-topics for both Labels and tags, I have added the reltive steps accordingly.

Reviewer 2: This content should be placed where the user is trying to add tags

Author: I will add the steps accordingly.

Author: Have added the steps here in "installing-gcp-cluster-creation.adoc" module where the cluster is created.

** An `OrganizationID` must consist of decimal numbers, and cannot have leading zeros.
** A `ProjectID` must be 6 to 30 characters in length, can only contain lowercase letters, numbers, and hyphens, and must start with a letter, and cannot end with a hyphen.
* A tag key and value must have a minimum of 1 character and can have a maximum of 63 characters.
* A tag key must contain only uppercase and lowercase alphanumeric characters, hyphen (`-`), underscore (`_`), and period (`.`).
* A tag value must contain only uppercase and lowercase alphanumeric characters, hyphen (`-`), underscore (`_`), period (`.`), at sign (`@`), percent sign (`%`), equals sign (`=`), plusse (`+`), colon (`:`), comma (`,`), asterisk (`*`), pound sign (`$`), ampersand (`&`), parentheses (`()`), square braces (`[]`), curly braces (`{}`), and space.
* A tag key and value must begin and end with an alphanumeric character.
* Tag value must be one of the pre-defined values for the key.
* You can configure a maximum of 50 tags.
* There should be no tag key with the same value as any of the existing tag keys on a resource, inherited from the parent resource.

> .Cluster creation phase

Author: Acknowledged. I will add this as a module and add the following steps here accordingly.

[NOTE]
====
`Labels` and `tags` are defined during the `install-config.yaml` creation phase, and cannot be modified or updated with new `labels` and `tags` after cluster creation.
====

:FeatureName: Support for user-defined `labels` and `tags` for GCP
include::snippets/technology-preview.adoc[]

The installation program requires that a service account includes a `TagUser` role, so that the program can create the {product-title} cluster with defined tags at both organization and project levels.

You can use the `.platform.gcp.userLabels` and `platform.gcp.userTags` fields to define the list of desired labels and tags in `install-config.yaml` file. The `.platform.gcp.userLabels` and `.platform.gcp.userTags` fields are a YAML array of objects type.

> .Sample `install-config.yaml` file

Author: Ack.

[source,yaml]
----
apiVersion: v1
featureSet: TechPreviewNoUpgrade
platform:
 gcp:
   userLabels:<1>
   - key: <label_key>
     value: <label_value>
   userTags:<2>
   - parentID: <OrganizationID/ProjectID><3>
     key: <tag_key_short_name>
     value: <tag_value_short_name>
----

> <1> `userLabels`: Adds keys and values as labels to the resources created on GCP.

Author: Ack.

<2> `userTags`: Adds keys and values as tags to the resources created on GCP.
<3> `parentID`: The ID of the hierarchical resource where the tags are defined, at the organization or the project level.

.Limitations for the user-defined tags

* Tags will not be attached to the following items:
** Control plane instances and storage buckets created by the installation program
** Compute instances created by the Machine API provider for GCP
** Filestore instance resources created by the GCP Filestore CSI Driver Operator
** Compute disk and compute image resources created by the GCP PD CSI Driver Operator
* Tags are not supported for buckets located in the following regions:
** `us-east2` 
** `us-east3`
* Cluster Image Registry Operator does not throw any error but skips processing tags when the buckets are created in the tags unsupported region.
* Tags must not be restricted to particular service accounts, because operators create and use service accounts with minimal roles.

> * {product-title} does not create any tag's key and value resources.

Author: Ack. Let me rephrase it.

* {product-title} specific tags are not added to any resource.

> * Labels for `Compute Address` are supported in the GCP beta version. {product-title} does not add labels to the resource.

Author: @bharath-b-rh Can you take a look at this?

> * Labels for `Compute Address` are supported in the GCP beta version. {product-title} does not add labels to the resource.

Reviewer 3: GCP docs uses it in both ways, `ComputeAddress` when referring to the object kind and `Compute Address` in general reference to the object.

> * Labels for `Compute Address` are supported in the GCP beta version. {product-title} does not add labels to the resource.

Author: Ack.

> .Configuration of labels and tags for GCP after the cluster creation

Author: Ack.

After creating the {product-title} cluster, you can access the list of the labels and tags defined for the GCP resources by referring to the new fields, `.status.platformStatus.gcp.resourceLabels` and `.status.platformStatus.gcp.resourceTags` present in the `infrastructures.config.openshift.io/cluster` object.

> .Sample `infrastructure.yaml` file

Author: Done.

[source,yaml]
----
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
 name: cluster
spec:
 platformSpec:
   type: GCP
status:
 infrastructureName: <cluster_id>
 platform: GCP
 platformStatus:
   gcp:
     resourceLabels:
     - key: <label_key>
       value: <label_value>
     resourceTags:
     - key: <tag_key_short_name>
       parentID: <OrganizationID/ProjectID>
       value: <tag_value_short_name>
   type: GCP
----

> Along with the user-defined labels, resources have a label defined by the {product-title}. The format of the {product-title} labels is `kubernetes-io-cluster-<cluster_id>:owned`. The `cluster_id` parameter has the value of `.status.infrastructureName` present in `infrastructures.config.openshift.io/cluster`.

Author: Ack.

> User-defined labels and {product-title} labels can be found on the following GCP resources:

Author: Ack.

* Compute Instance
* Compute Disk
* Compute Forwarding Rule
* Compute Image
* DNS Managed Zone
* Filestore Instance
* Storage Bucket

> User-defined tags can be found added to the GCP resources, that is, storage bucket.

Author: It was suggested from the SME. Let me still see if we can rephrase it.

[role="_additional-resources"]
.Additional resources

> * For more information about identifying the `OrganizationID` or `ProjectID`, see:

Author: Done.

** link:https://cloud.google.com/resource-manager/docs/creating-managing-organization#retrieving_your_organization_id[OrganizationID]
** link:https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects[ProjectID]

> * For more information about tags and labels, see link:https://cloud.google.com/storage/docs/tags-and-labels#tags[Tags and labels].

Author: Okay

// Module included in the following assemblies:
// * installing/installing_gcp/installing-gcp-customizations.adoc

:_content-type: CONCEPT
[id="installing-gcp-user-defined-labels-and-tags_{context}"]

> = User-defined labels and tags for GCP

Reviewer 2: I know I suggested this. But looking at the topic hierarchy, I think we should rename to something like below:

```suggestion
= Managing user-defined labels and tags for GCP
```

Author: Ack.

// Module included in the following assemblies:
// * installing/installing_gcp/installing-gcp-customizations.adoc

:_content-type: CONCEPT
[id="installing-gcp-user-defined-labels-and-tags_{context}"]
= User-defined labels and tags for GCP

Google Cloud Platform (GCP) provides labels and tags that help to identify and organize the resources created for a specific {product-title} cluster, making them easier to manage. 

You can define labels and tags for each GCP resource only during {product-title} cluster installation. 

[IMPORTANT]
====
User-defined labels and tags are not supported for {product-title} clusters upgraded to {product-title} 4.14 version.
====

.User-defined labels

> User-defined labels and {product-title} specific labels are applied only to resources created by {product-title} installer and its core components such as: 

Reviewer 2: ```suggestion
User-defined labels and {product-title} specific labels are applied only to resources created by {product-title} installation program and its core components such as: 
```

Author: Ack.

* GCP filestore CSI Driver Operator
* GCP PD CSI Driver Operator
* Image Registry Operator
* Machine API provider for GCP

> The user-defined labels and {product-title} specific labels are not applied on the resources created by any other operators or the kubernetes in-tree components that create resources, for example, the Ingress load balancers.

Reviewer 2: I think this is also a limitation. Kindly confirm.

Author: This was more like intro but the info was added in the limitations and then to avoid duplication of the data, I had removed it. Let me change it.

Author: Confirmed from the SME "It's not a limitation, since those components are not owned by openshift, it's out of scope."
